### PR TITLE
[BD-24] [TNL-7318]: BB-2355 - Add decoding and scope verification methods

### PR DESCRIPTION
**Pull request by giovannicimolin**
This PR adds a few token verification mechanisms to support LTI Advantage services.
This is PR 2 of https://github.com/edx/xblock-lti-consumer/pull/92.

**Testing instructions:**
Currently there are no endpoints using this, so just make sure the code is same and checks are passing.

**Reviewer:** 
- [ ] @viadanna 
- [ ] @davidjoy 
- [ ] @davestgermain

### CODE REVIEW

**Comment by edx-webhook** [comment]
Thanks for the pull request, @giovannicimolin! I've created [BLENDED-527](https://openedx.atlassian.net/browse/BLENDED-527) to keep track of it in Jira.   More details are on the [BD-24](https://openedx.atlassian.net/wiki/spaces/COMM/pages/1545076784/BD-24+LTI+v1.3+Improvements+Advantage+Support) project page.

When this pull request is ready, tag your edX technical lead.

File: lti_consumer/lti_1p3/consumer.py [hunk 1]
```diff
@@ -397,3 +397,29 @@ def _validate_preflight_response(self, response):
             assert response.get("redirect_uri") == self.launch_url
         except AssertionError:
             raise exceptions.PreflightRequestValidationFailure()
+
+    def check_token(self, token, allowed_scopes):
+        """
+        Check if token has access to allowed scopes.
+        """
+        if not allowed_scopes:
```
Comment (viadanna) [inline lti_consumer/lti_1p3/consumer.py:405]: nit: can this be merged with the last if?

File: lti_consumer/lti_1p3/consumer.py [hunk 2]
```diff
@@ -397,3 +397,29 @@ def _validate_preflight_response(self, response):
...
+            token,
+            # The issuer of the token is the platform
+            iss=self.iss,
+        )
+        # Tokens are space separated
+        token_scopes = token_contents['scopes'].split(' ')
+
+        # Check if token has permission for the requested scope,
+        # and throws exception if not.
+        # In `allowed_scopes` is empty, return true (just check
```
Comment (viadanna) [inline lti_consumer/lti_1p3/consumer.py:418]: typo: In/If

File: lti_consumer/lti_1p3/key_handlers.py [hunk 3]
```diff
@@ -182,3 +182,34 @@ def get_public_jwk(self):
...
+        try:
+            # Verify message signature
+            message = JWS().verify_compact(token, keys=[self.key])
+
+            # If message is valid, check expiration from JWT
+            if 'exp' in message and message['exp'] < time.time():
+                raise exceptions.TokenSignatureExpired()
+
+            # Validate issuer claim
+            if iss and ('iss' not in message or message['iss'] != iss):
```
Comment (viadanna) [inline lti_consumer/lti_1p3/key_handlers.py:202]: Do you think nesting ifs would make this more readable?

File: lti_consumer/lti_1p3/key_handlers.py [hunk 4]
```diff
@@ -182,3 +182,34 @@ def get_public_jwk(self):
...
+                raise exceptions.InvalidClaimValue()
+
+            # Validate audience claim
+            if aud and ('aud' not in message or aud not in message['aud']):
```
Comment (viadanna) [inline lti_consumer/lti_1p3/key_handlers.py:206]: Same as above.

**Pull request merged**